bokehjs/src/coffee/core/util/serialization.ts

```diff
@@ -1,215 +1,260 @@
     } else if (arr.BYTES_PER_ELEMENT === 4) {
-      swap32(arr);
+      swap32(arr)
     } else if (arr.BYTES_PER_ELEMENT === 8) {
-      swap64(arr);
+      swap64(arr)
     }
   }
-  return [arr, shape];
-};
-
-export const process_array = function(obj, buffers) {
-  if (isObject(obj) && '__ndarray__' in obj) {
-    return decode_base64(obj);
-  } else if (isObject(obj) && '__buffer__' in obj) {
-    return process_buffer(obj, buffers);
-  } else if (isArray(obj)) {
-    return [obj, []];
-  }
-};
+  return [arr, shape]
+}
+
+export function process_array(obj: NDArray | BufferSpec | ArrayLike<any>, buffers: [any, any][]): [ArrayLike<any>, number[]] {
+  if (isObject(obj) && '__ndarray__' in obj)
+    return decode_base64(obj)
+  else if (isObject(obj) && '__buffer__' in obj)
+    return process_buffer(obj, buffers)
+  else if (isArray(obj))
+    return [obj, []]
+  else
+    return undefined as never
+}
 
-export const arrayBufferToBase64 = function(buffer) {
-  const bytes = new Uint8Array(buffer);
+export function arrayBufferToBase64(buffer: ArrayBuffer): string {
+  const bytes = new Uint8Array(buffer)
   const binary = map(bytes, (b) => String.fromCharCode(b))
-  return btoa(binary.join(""));
-};
+  return btoa(binary.join(""))
+}
 
-export const base64ToArrayBuffer = function(base64) {
-  const binary_string = atob(base64);
+export function base64ToArrayBuffer(base64: string): ArrayBuffer {
+  const binary_string = atob(base64)
   const len = binary_string.length
-  const bytes = new Uint8Array(len);
+  const bytes = new Uint8Array(len)
   for (let i = 0, end = len; i < end; i++) {
-    bytes[i] = binary_string.charCodeAt(i);
-  }
-  return bytes.buffer;
-};
-
-export const decode_base64 = function(input) {
-  let array;
-  const bytes = base64ToArrayBuffer(input['__ndarray__']);
-  const dtype = input['dtype'];
-  if (dtype in ARRAY_TYPES) {
-    array = new (ARRAY_TYPES[dtype])(bytes);
+    bytes[i] = binary_string.charCodeAt(i)
   }
-  const shape = input['shape'];
-  return [array, shape];
-};
+  return bytes.buffer
+}
+
+export interface NDArray {
+  __ndarray__: string
+  shape?: Shape
+  dtype: DType
+}
+
+export function decode_base64(input: NDArray): [TypedArray, Shape] {
+  const bytes = base64ToArrayBuffer(input.__ndarray__)
+  const dtype = input.dtype
+  const shape = input.shape!
+
+  let array: TypedArray
+  if (dtype in ARRAY_TYPES)
+    array = new (ARRAY_TYPES[dtype])(bytes)
+  else
+    throw new Error(`unknown dtype: ${dtype}`)
+
+  return [array, shape]
+}
+
+export function encode_base64(array: TypedArray, shape?: Shape): NDArray {
+  const b64 = arrayBufferToBase64(array.buffer)
+  const name = array.constructor.name
+
+  let dtype: DType
+  if (name in DTYPES)
+    dtype = DTYPES[name as ArrayName]
+  else
+    throw new Error(`unknown array type: ${name}`)
 
-export const encode_base64 = function(array, shape) {
-  const b64 = arrayBufferToBase64(array.buffer);
-  const dtype = DTYPES[array.constructor.name];
```

Comment:
How this didn't fail before in phantomjs? `TypedArray.name` is not supported in phantomjs. Currently this fails with:
```
[JS] Error: unknown array type: undefined
[JS] [bokeh] Failed to repull session Error: unknown array type: undefined
```
in all server examples.

Reply:
I have no idea offhand, @canavandl any thoughts?

Reply:
I have no idea. Sorry